[FEATURE] Ne pas recalculer l'acces aux certifications complementaires dans les handlers (PIX-3941)

aceol opened the pull request:
## :christmas_tree: Problème
On calcule à plusieurs niveaux si le candidat est éligible aux certifications complémentaires. Ce n'est pas optimal et peut être faussé si les données ont changé entre temps

## :gift: Solution
La PR PIX-3689 enregistre au démarrage d'une certification, le démarrage des certifications complementaires. On peut donc réutiliser cette info pour éviter les recalculs dans les handlers

## :star2: Remarques
Pour ajouter le badge cléA a pixCertifDroit:
`update "badge-acquisitions" set "userId"=111 where "badgeId" = 118;`

## :santa: Pour tester
Lancer une certification avec Pix+droit, et cléA (certifdroit@example.net)
Verifier que l'obtention des certifications complémentaires est toujours OK.

Lancer une certification avec Pix+droit, et cléA. 
Avant de terminer la certif, supprimer les lignes de la table `complementary-certification-courses`.
Verifier que l'obtention des certifications complémentaires est KO.

aceol committed:
feat(api) Check if certification course has started pix+ complementary certification

aceol committed:
feat(api) Check if certification course has started clea complementary certification

aceol committed:
feat(api) Check if certification course has started clea complementary certification

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr3783.review.pix.fr
- App (.org): https://app-pr3783.review.pix.org
- Orga: https://orga-pr3783.review.pix.fr
- Certif: https://certif-pr3783.review.pix.fr
- Admin: https://admin-pr3783.review.pix.fr
- API: https://api-pr3783.review.pix.fr/api/

 Please check it out!

MathieuGilet commented on api/tests/unit/domain/events/handle-pix-plus-certifications-scoring_test.js:
```diff
@@ -1,33 +1,25 @@
-    // eslint-disable-next-line mocha/no-setup-in-describe
-    getByCertificationCourseId: _.noop(),
-  };
+  const certificationAssessmentRepository = {};
```
Perso j'aime bien juste déclarer les repo ici, sans les instancier.

aceol replied:
Si on les instancie pas, on ne peut pas utiliser le `injectables` (on perdrait la ref)

MathieuGilet commented on api/lib/domain/events/handle-clea-certification-rescoring.js:
```diff
@@ -1,23 +1,22 @@
-  }
-  const cleaCertificationResult = await cleaCertificationResultRepository.get({ certificationCourseId });
-  if (!cleaCertificationResult.isTaken()) {
+  const hasRunCla = await complementaryCertificationCourseRepository.hasComplementaryCertification({
```
`hasRunClea`

MathieuGilet commented on api/tests/unit/domain/events/handle-certification-scoring_test.js:
```diff
@@ -202,6 +221,12 @@ describe('Unit | Domain | Events | handle-certification-scoring', function () {
           .withArgs(certificationAssessment.certificationCourseId)
           .resolves(certificationCourse);
         certificationCourseRepository.update.resolves(certificationCourse);
+        complementaryCertificationCourseRepository.hasComplementaryCertification
```
Vu qu'on est dans un contete 'when scoring is successful', je ne comprends pas pourquoi le repo retourne false

aceol replied:
En fait c'est un loupé, ce repo n'est pas utilisé ici, ca n'a pas d'impact (mais il faut le degager)

MathieuGilet commented on api/tests/unit/domain/events/handle-clea-certification-rescoring_test.js:
```diff
@@ -108,36 +102,5 @@ describe('Unit | Domain | Events | handle-clea-certification-rescoring', functio
         });
       });
     });
-
```
J'ai l'impression qu'il manque un test montrant que quand le hasComplementaryCertification renvoie false, la certif cléa ne doit pas être rescoré

aceol replied:
Ca devrait correspondre à cela (deja dans le code)!?
```
 context('when CleA certification was not even taken in the first place', function () {
      it('should not build or save no partner certification scoring', async function () {
```

MathieuGilet commented on api/tests/unit/domain/events/handle-clea-certification-scoring_test.js:
```diff
@@ -55,69 +54,15 @@ describe('Unit | Domain | Events | handle-clea-certification-scoring', function
     expect(error).not.to.be.null;
   });
 
-  context('when badge is not acquired', function () {
```
On est d'accord que ce cas est géré par le fait d'ajouter ou non une certif complémentaire au certif course?

aceol replied:
Oui c'est ca. On n'a plus à verifier le badge ici. C'est fait une seul fois, à la creation du `certificationCourse`

MathieuGilet commented on api/tests/unit/domain/events/handle-clea-certification-scoring_test.js:
```diff
@@ -174,16 +107,23 @@ describe('Unit | Domain | Events | handle-clea-certification-scoring', function
           .withArgs({ knowledgeElements, targetProfile, badge })
           .returns(true);
 
+        complementaryCertificationCourseRepository.hasComplementaryCertification
```
**Suggestion**: Placer les stubs dans l'ordre des appels dans le code

aceol replied:
Oui ca pourrait simplifier la comprehension

MathieuGilet commented on api/lib/domain/events/handle-clea-certification-rescoring.js:
```diff
@@ -1,23 +1,22 @@
   checkEventTypes(event, eventTypes);
   const { certificationCourseId } = event;
 
-  const certificationCenter = await certificationCenterRepository.getByCertificationCourseId(certificationCourseId);
```
Cette méthode de repo ne me semble plus utilisée suite à ce commit, on la supprime ?

aceol reviewed (commented):
J'ai l'impression qu'il y a du code qui ne devrait pas etre là. Je vais investiguer

clemlatz commented on api/tests/integration/infrastructure/repositories/complementary-certification-course-repository_test.js:
```diff
@@ -0,0 +1,93 @@
+describe('Integration | Repository | complementary-certification-courses-repository', function () {
+  describe('#hasComplementaryCertification', function () {
+    describe('when complementary certification has been started for the certification course', function () {
+      it('return true', async function () {
```
```suggestion
      it('returns true', async function () {
```

clemlatz commented on api/tests/integration/infrastructure/repositories/complementary-certification-course-repository_test.js:
```diff
@@ -0,0 +1,93 @@
+    });
+
+    describe('when another complementary certification has been started for the certification course', function () {
+      it('return false', async function () {
```
```suggestion
      it('returns false', async function () {
```

clemlatz commented on api/tests/integration/infrastructure/repositories/complementary-certification-course-repository_test.js:
```diff
@@ -0,0 +1,93 @@
+    });
+
+    describe('when complementary certification has been started for another certification course', function () {
+      it('return false', async function () {
```
```suggestion
      it('returns false', async function () {
```

clemlatz commented on api/lib/domain/events/handle-clea-certification-rescoring.js:
```diff
@@ -1,23 +1,22 @@
+    certificationCourseId,
+    complementaryCertificationName: CLEA,
+  });
+  if (!hasRunCla) {
```
```suggestion
  if (!hasRunClea) {
```

aceol committed:
refacto(api) Remove duplicate code

Check for cleA badge avalaibility is now redondant. If we remove it from scoring. Rescoring specific code can be safely removed

aceol commented:
TODO mettre sous toggle `FT_IS_COMPLEMENTARY_CERTIFICATION_SUBSCRIPTION_ENABLED`

octo-topi reviewed (changes requested):
Prendre en compte le FT dans tous les cas
CRéation d'une nouvelle branche   `pix-3941-use-complementary-certification-course-in-scoring` et d'une nouvelle PR

octo-topi commented:
Remplacé par https://github.com/1024pix/pix/pull/3800
